packages/flutter/lib/src/widgets/drag_boundary.dart

```diff
@@ -0,0 +1,91 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:ui';
+
+import 'framework.dart';
+
+/// The interface for defining the algorithm that a boundary handles the specified type of shape to be dragged within.
+///
+/// See also:
+///  * [DragBoundary], which uses this class.
+///
+/// `T` is a data class that defines the shape being dragged. For example, when dragging a rectangle within the boundary, `T` should be a `Rect`. 
+abstract class DragBoundaryDelegate<T> {
+  /// Returns whether the specified drag object position is within the boundary.
+  bool isWithinBoundary(T position);
+
+  /// Returns the position of the given drag object after moving it fully inside
+  /// the boundary with the shortest distance.
+  T nearestPositionWithinBoundary(T position);
+}
+
+class _DragBoundaryDelegateForRect extends DragBoundaryDelegate<Rect> {
+  _DragBoundaryDelegateForRect(this.boundary);
+  final Rect boundary;
+  @override
+  bool isWithinBoundary(Rect position) {
+    return boundary.contains(position.topLeft) && boundary.contains(position.bottomRight);
+  }
+
+  @override
+  Rect nearestPositionWithinBoundary(Rect position) {
+    assert(() {
```

Comment:
`assert` should only be used for violation of the assumption of the Flutter library; in other words, if Flutter has a bug. This statement is instead a bug of the user code, and should either return null or throw exceptions (usually `FlutterError`.